system/jlib/jfile.cpp

```diff
@@ -7106,3 +7111,244 @@ const FileSystemProperties & queryFileSystemProperties(const char * filename)
+
+        /* NB: in this impl. inotify_event flags are same as FileWatchEvents bit masks
+         * If other implementations differ, events will need mapping to the underlying impl.
+         */
+        int iNotifyEvents = static_cast<int>(events);
+        int wd = inotify_add_watch(inotifyFd, filename, iNotifyEvents);
+        if (-1 == wd)
+            throw makeErrnoExceptionV(-1, "CFileEventWatcher::add() - Cannot watch '%s'", filename);
+        Owned<CMonitoredItem> item = new CMonitoredItem(filename, wd, events);
+        monitoredFiles.emplace(filename, item);
+        monitoredFilesByFd.emplace(wd, item);
+
+        return true;
+    }
+    virtual bool remove(const char *filename) override
+    {
+        if (isEmptyString(filename))
+            return false;
+
+        CriticalBlock b(crit);
+        auto iter = monitoredFiles.find(filename);
+        if (iter == monitoredFiles.end()) // filename is not being monitored
+            return false;
+        CMonitoredItem *item = iter->second;
+        if (item->pendingRemoval) // already flagged for removal
+            return false;
+
+        item->pendingRemoval = true;
+        pendingRemovals.emplace(item->watchFd);
+
+        if (-1 == inotify_rm_watch(inotifyFd, item->watchFd))
+            throw makeErrnoException("CFileEventWatcher::remove() - inotify_rm_watch()");
+        return true;
+    }
+// IThreaded impl.    
+    virtual void threadmain() override
+    {
+        try
+        {
+            while (true)
+            {
+                int poll_num = poll(toPoll, 2, -1);
+                if (-1 == poll_num)
+                {
+                    if (EINTR == errno)
+                        continue;
+                    throw makeErrnoException("CFileEventWatcher - poll error");
+                }
+
+                if (poll_num)
+                {
+                    if (toPoll[0].revents & POLLIN)
+                    {
+                        if (stopped)
+                            break;
+                    }
+                    else if (toPoll[1].revents & POLLIN)
+                    {
+                        handle_events(inotifyFd);
+
+                        // clear any pending deletes
+                        CriticalBlock b(crit);
+                        for (const auto fd: pendingRemovals)
+                        {
+                            auto iter = monitoredFilesByFd.find(fd);
+                            if (iter != monitoredFilesByFd.end())
+                            {
+                                CMonitoredItem *item = iter->second;
+                                if (item->pendingRemoval) // else fd has been reused/readded
+                                {
+                                    monitoredFiles.erase(item->filename);
+                                    monitoredFilesByFd.erase(iter);
+                                }
+                            }
+                        }
+                    }
```

Comment:
Should clear pendingRemovals (outside of cs)

Reply:
I forgot to clear them.. I'll change,
but I need to do within the cs, others could be adding/removing and updating pendingRemovals on other threads.

Reply:
good point